Panic with pointer-to-interface field

Example is a little artificial though - don't think it's common to find pointers to interfaces in actual production code, but still might occasionally happen I'm afraid.

Sample code demonstrating issue:
```
package main

import (
	"testing"

	"github.com/mitchellh/copystructure"
	"github.com/stretchr/testify/require"
)

func TestCopyWithIfacePtr(t *testing.T) {
	type Nested struct {
		Field string
	}

	type External struct {
		IfacePtr *interface{}
	}

	ifacePtr := func(v interface{}) *interface{} {
		return &v
	}

	copy, err := copystructure.Copy(External{IfacePtr: ifacePtr(Nested{Field: "111"})})
	require.Nil(t, err)
	require.Equal(t, External{IfacePtr: ifacePtr(Nested{Field: "111"})}, copy)
}
```

Output:
```
--- FAIL: TestCopyWithIfacePtr (0.00s)
panic: reflect.Set: value of type *main.Nested is not assignable to type *interface {} [recovered]
	panic: reflect.Set: value of type *main.Nested is not assignable to type *interface {}

goroutine 5 [running]:
panic(0x63a940, 0xc420013650)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
testing.tRunner.func1(0xc420090180)
	/usr/local/go/src/testing/testing.go:579 +0x25d
panic(0x63a940, 0xc420013650)
	/usr/local/go/src/runtime/panic.go:458 +0x243
reflect.Value.assignTo(0x62c120, 0xc420013610, 0x16, 0x697462, 0xb, 0x62e220, 0x0, 0x1, 0x1, 0x0)
	/usr/local/go/src/reflect/value.go:2163 +0x35c
reflect.Value.Set(0x62e220, 0xc420032058, 0x196, 0x62c120, 0xc420013610, 0x16)
	/usr/local/go/src/reflect/value.go:1333 +0xa4
github.com/mitchellh/copystructure.(*walker).Exit(0xc420098090, 0x7, 0x0, 0x0)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/copystructure/copystructure.go:202 +0x86a
github.com/mitchellh/reflectwalk.walkStruct(0x656800, 0xc4200135b0, 0x19, 0x687840, 0xc420098090, 0x0, 0x10)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/reflectwalk/reflectwalk.go:330 +0x40c
github.com/mitchellh/reflectwalk.walk(0x656800, 0xc4200135b0, 0x19, 0x687840, 0xc420098090, 0x0, 0x0)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/reflectwalk/reflectwalk.go:180 +0x4b2
github.com/mitchellh/reflectwalk.Walk(0x656800, 0xc4200135b0, 0x687840, 0xc420098090, 0x0, 0xc42002fec8)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/reflectwalk/reflectwalk.go:84 +0x155
github.com/mitchellh/copystructure.Config.Copy(0x647c00, 0x0, 0x656800, 0xc4200135b0, 0x4764dd, 0x647c80, 0xc4200135b0, 0xc4200135b0)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/copystructure/copystructure.go:69 +0x81
github.com/mitchellh/copystructure.Copy(0x656800, 0xc4200135b0, 0xc4200135b0, 0x656880, 0xc4200135a0, 0xf3)
	/home/sbinq/work/gopath/vendor/src/github.com/mitchellh/copystructure/copystructure.go:13 +0x4c
_/home/sbinq/work/gopath/src.TestCopyWithIfacePtr(0xc420090180)
	/home/sbinq/work/gopath/src/sample_test.go:23 +0xab
testing.tRunner(0xc420090180, 0x6b6e10)
	/usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x2ec
exit status 2
FAIL	_/home/sbinq/work/gopath/src	0.005s
```

--- a/copystructure.go
+++ b/copystructure.go
@@ -405,6 +405,23 @@ func (w *walker) replacePointerMaybe() {
 	}
 
 	v := w.valPop()
+
+	// If the expected type is a pointer to an interface of any depth,
+	// such as *interface{}, **interface{}, etc., then we need to convert
+	// the value "v" from *CONCRETE to *interface{} so types match for
+	// Set.
+	//
+	// Example if v is type *Foo where Foo is a struct, v would become
+	// *interface{} instead. This only happens if we have an interface expectation
+	// at this depth.
+	//
+	// For more info, see GH-16
+	if iType, ok := w.ifaceTypes[ifaceKey(w.ps[w.depth], w.depth)]; ok && iType.Kind() == reflect.Interface {
+		y := reflect.New(iType)           // Create *interface{}
+		y.Elem().Set(reflect.Indirect(v)) // Assign "Foo" to interface{} (dereferenced)
+		v = y                             // v is now typed *interface{} (where *v = Foo)
+	}
+
 	for i := 1; i < w.ps[w.depth]; i++ {
 		if iType, ok := w.ifaceTypes[ifaceKey(w.ps[w.depth]-i, w.depth)]; ok {
 			iface := reflect.New(iType).Elem()

--- a/copystructure_test.go
+++ b/copystructure_test.go
@@ -24,6 +24,32 @@ func TestCopy_complex(t *testing.T) {
 	}
 }
 
+func TestCopy_interfacePointer(t *testing.T) {
+	type Nested struct {
+		Field string
+	}
+
+	type Test struct {
+		Value *interface{}
+	}
+
+	ifacePtr := func(v interface{}) *interface{} {
+		return &v
+	}
+
+	v := Test{
+		Value: ifacePtr(Nested{Field: "111"}),
+	}
+	result, err := Copy(v)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	if !reflect.DeepEqual(result, v) {
+		t.Fatalf("bad: %#v", result)
+	}
+}
+
 func TestCopy_primitive(t *testing.T) {
 	cases := []interface{}{
 		42,

